### PR TITLE
Add a Sabre plugin that emits an event

### PR DESCRIPTION
**Pull request by nickvergessen**
@DeepDiver1975 @PVince81

### CODE REVIEW

File: lib/private/connector/sabre/listenerplugin.php [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+	}
+
+	/**
+	 * This method is called before any HTTP method and returns http status code 503
+	 * in case the system is in maintenance mode.
+	 *
+	 * @return bool
+	 */
+	public function emitListener() {
+		$this->dispatcher->dispatch('OC\Connector\Sabre::dav');
```
Comment (DeepDiver1975) [inline lib/private/connector/sabre/listenerplugin.php]: let's add the event to it as well

```php
$this->dispatcher->dispatch('OC\Connector\Sabre::beforeMethod');

Comment (DeepDiver1975) [inline lib/private/connector/sabre/listenerplugin.php]: one day we might want to listen to other sabre events as well

Comment (nickvergessen) [inline lib/private/connector/sabre/listenerplugin.php]: fixed

**Comment by DeepDiver1975** [comment]
````
PHP Fatal error:  Cannot use Sabre\DAV\Server as Server because the name is already in use in /var/jenkins/workspace/core-ci-linux@6/database/sqlite/label/SLAVE/lib/private/connector/sabre/listenerplugin.php on line 24
08:34:38 PHP Stack trace:
08:34:38 PHP   1. {main}() /usr/local/bin/phpunit:0
08:34:38 PHP   2. PHPUnit_TextUI_Command::main() /usr/local/bin/phpunit:537
08:34:38 PHP   3. PHPUnit_TextUI_Command->run() phar:///usr/local/bin/phpunit/phpunit/TextUI/Command.php:105
08:34:38 PHP   4. PHPUnit_TextUI_TestRunner->doRun() phar:///usr/local/bin/phpunit/phpunit/TextUI/Command.php:153
08:34:38 PHP   5. PHPUnit_Framework_TestSuite->run() phar:///usr/local/bin/phpunit/phpunit/TextUI/TestRunner.php:406
08:34:38 PHP   6. PHPUnit_Framework_TestSuite->run() phar:///usr/local/bin/phpunit/phpunit/Framework/TestSuite.php:722
08:34:38 PHP   7. PHPUnit_Framework_TestCase->run() phar:///usr/local/bin/phpunit/phpunit/Framework/TestSuite.php:722
08:34:38 PHP   8. PHPUnit_Framework_TestResult->run() phar:///usr/local/bin/phpunit/phpunit/Framework/TestCase.php:699
08:34:38 PHP   9. PHP_Invoker->invoke() phar:///usr/local/bin/phpunit/phpunit/Framework/TestResult.php:607
08:34:38 PHP  10. call_user_func_array:{phar:///usr/local/bin/phpunit/php-invoker/Invoker.php:93}() phar:///usr/local/bin/phpunit/php-invoker/Invoker.php:93
08:34:38 PHP  11. PHPUnit_Framework_TestCase->runBare() phar:///usr/local/bin/phpunit/php-invoker/Invoker.php:93
08:34:38 PHP  12. PHPUnit_Framework_TestCase->runTest() phar:///usr/local/bin/phpunit/phpunit/Framework/TestCase.php:743
08:34:38 PHP  13. ReflectionMethod->invokeArgs() phar:///usr/local/bin/phpunit/phpunit/Framework/TestCase.php:866
08:34:38 PHP  14. Test\Connector\Sabre\RequestTest\UploadTest->testBasicUpload() phar:///usr/local/bin/phpunit/phpunit/Framework/TestCase.php:866
08:34:38 PHP  15. Test\Connector\Sabre\RequestTest\RequestTest->request() /var/jenkins/workspace/core-ci-linux@6/database/sqlite/label/SLAVE/tests/lib/connector/sabre/requesttest/uploadtest.php:17
08:34:38 PHP  16. Test\Connector\Sabre\RequestTest\RequestTest->getSabreServer() /var/jenkins/workspace/core-ci-linux@6/database/sqlite/label/SLAVE/tests/lib/connector/sabre/requesttest/requesttest.php:76
08:34:38 PHP  17. OC\Connector\Sabre\ServerFactory->createServer() /var/jenkins/workspace/core-ci-linux@6/database/sqlite/label/SLAVE/tests/lib/connector/sabre/requesttest/requesttest.php:123
08:34:38 PHP  18. spl_autoload_call() /var/jenkins/workspace/core-ci-linux@6/database/sqlite/label/SLAVE/tests/lib/connector/sabre/requesttest/requesttest.php:77
08:34:38 PHP  19. OC\Autoloader->load() /var/jenkins/workspace/core-ci-linux@6/database/sqlite/label/SLAVE/tests/lib/connector/sabre/requesttest/requesttest.php:0

**Comment by nickvergessen** [comment]
fixed

File: lib/private/connector/sabre/serverfactory.php [hunk 2]
```diff
@@ -71,6 +74,7 @@ public function createServer($baseUri, $requestUri, BackendInterface $authBacken
 		$server->addPlugin(new \OC\Connector\Sabre\FilesPlugin($objectTree));
 		$server->addPlugin(new \OC\Connector\Sabre\ExceptionLoggerPlugin('webdav', $this->logger));
 		$server->addPlugin(new \OC\Connector\Sabre\LockPlugin($objectTree));
+		$server->addPlugin(new \OC\Connector\Sabre\ListenerPlugin($this->dispatcher));
```
Comment (PVince81) [inline lib/private/connector/sabre/serverfactory.php]: Something to think about: should the plugin also be registered for public webdav ?
If yes, should the plugin pass an additional argument to the callback to tell the listener that it's public webdav ?

Comment (PVince81) [inline lib/private/connector/sabre/serverfactory.php]: if not then the listener can check the incognito mode value.

**Comment by PVince81** [comment]
:+1: and please check my comment

**Comment by PVince81** [comment]
@DeepDiver1975 second thumbs up + merge ?

**Comment by DeepDiver1975** [comment]
:+1:

**Pull request merged**